First draft of the "InnerSource Activity Score" pattern (#189)
gruetter/innerSourceTest: innersource-activity-score.md

## Title 

> InnerSource Activity Score

Reviewer: Two ideas around the title:

- the pattern is about **project** activity it seems? As there could be other similar scores derived from this in the future (e.g. contributor activity score, team activity score, organization activity score, etc), maybe the title should indicate that this is about a repo or project specifically?
- all patterns are about InnerSource, so in theory all patterns could contain the world "InnerSource". :) For your pattern, does the word make a significant difference, or could you leave it away?

Reviewer: If you end up changing the title, you may also consider the filename.

Author: I think all stats are closely related to the codebase, so i renamed it to "repository activity score", great feedback, thx!

## Patlet

> The InnerSource activity score is a numeric value that represents the (GitHub) activity of an InnerSource project. It can be used to create a ranked list of projects for the [InnerSource portal](innersource-portal.md) so that potential contributors can easily find the most active projects. It is derived automatically from GitHub repository statistics and may be enriched with KPIs from other tools or manual evaluations. 

Reviewer: From the template:
> Concise 1-2 sentence description of the problem and solution. Readers may quickly review dozens of these patlets to discover and browse the larger library of patterns. 

Squeezing everything into two sentences is really tough but still worth trying for the benefit of the readers :) Even just stating the problem first, and then following with the solution can already provide a lot of value.

Let me try:

```suggestion
Potential contributors want to find active InnerSource projects in need of their help. By calculating an InnerSource activity score for each project, a ranked list of projects can be created (e.g. on the [InnerSource portal](innersource-portal.md)), so that potential contributors can more easily determine which project they want to contribute to.
```

## Problem

Potential contributors want to find active projects with a lot of traction, but also fairly new and enthusiastic projects with a need for new contributions - easy and fast.

> **But in which order** shall InnerSource projects be presented? Typical ranking KPIs like *GitHub Stars*, *Number of Forks*, *Number of Commits*, *Lines of Code*,  *Last Update* aren't very accurate when it comes to activity of a project.

Reviewer 2: instead of "aren't very accurate", perhaps it is more appropriate to say "aren't sufficient to concisely indicate the activity of a project"

A new metric derived from several KPIs is needed to define a reliable and versatile score for a project's activity level.

> It can be used to sort or rank projects according to their activity level.

Reviewer: For brevity I would just use one of the words order/sort/rank to describe the main feature that the activity score enables.

```suggestion
It can be used to sort projects according to their activity level.
```

## Story

> When InnerSource is practiced for a long time or scales beyond a certain number of projects (let's say 50 to give a meaningful threshold) it is hard to find the currently most popular and active InnerSource projects. Projects that exist for a long are well-known but may no longer be very active. Fairly new projects on the other hand don't have a repuation or an active community yet.

Reviewer: ```suggestion
When InnerSource is practiced for a long time or scales beyond a certain number of projects (let's say 50 to give a meaningful threshold) it is hard to find the currently most popular and active InnerSource projects. Projects that exist for a long time are well-known but may no longer be very active. Fairly new projects on the other hand don't have a reputation or an active community yet.
```

A list of InnerSource projects should not be considered a static resource, but an exciting place to discover and explore new and active projects, just like a news page listing the most interesting topics of the day first. Thus it is beneficial when the order of the projects is regularly updated and changes according to the project's popularity and activity.   

These considerations let to a first prototype to calculate an InnerSource activity score, which worked surprisingly well and determines an ever-changing order of projects according to their activity.

> ## Context

Reviewer: I would add something to the context about the contribution culture of the developers in your situation.

The pattern suggests that the developers are "looking for projects to contribute to" at certain points in time.

Therefore I assume that something like this is true in you case:

> Contributors are free to commit a part of their time to InnerSource project. They may choose to contribute to a project that they depend on for the work in their regular team anyways. However they may also choose to contribute to something completely different, based on their interests and personal development goals.

Providing specifics about your context helps other readers to determine "Do I have this same particular situation?"

Author: added this part, thank you!

> Discovering InnerSource projects can be facilitated with the [InnerSource Portal](innersource-portal.md) and the [Gig Marketplace](gig-marketplace.md) pattern, or by promoting projects on other communication channels and platforms. The activity score defines a default order or ranking in which projects are presented to the community. 

Reviewer: For brevity I would just use one of the words order/sort/rank to describe the main feature that the activity score enables.

```suggestion
Discovering InnerSource projects can be facilitated with the [InnerSource Portal](innersource-portal.md) and the [Gig Marketplace](gig-marketplace.md) pattern, or by promoting projects on other communication channels and platforms. The activity score defines a default order in which projects are presented to the community. 
```

## Forces

> Automated KPIs that can be fetched by querying the GitHub API are only part of the truth. What about code quality, the availability of good documentation, or an active and helping community that makes the project a fun place to contribute.

Reviewer: ```suggestion
Automated KPIs that can be fetched by querying the GitHub API are only part of the truth. What about code quality, the availability of good documentation, or an active and helping community that makes the project a fun place to contribute?
```

Reviewer 2: Forces typically act upon the problem space in a way that is expressed through trade-offs. E.g., automated KPIs are easy to obtain but might not convey the whole picture. Soft KPIs are harder to gather but can add valuable meaning.

The "For now, we focus on a fully automated solution." isn't really a part of Forces but something that would appear in the Solution section (a decision on what to do with this particular Force).

Author: moved parts of it to solution, thx

    // new repos (created in the last 365 days) that have been updated recently (in the last 30 days) get a boost of up to 1000
    let iDaysSinceCreation = (new Date().getTime() - new Date(repo.created_at).getTime()) / 1000 / 86400;
    let iBoost = (1000 - Math.min(iDaysSinceLastUpdate, 100) * 10);
    iScore += (iDaysSinceCreation < 365 ? (iDaysSinceLastUpdate < 30 ? iBoost : 0) : 0);
    // give projects with contribution guidelines (CONTRIBUTING.md) file a static boost of 100
    iScore += (repo["_InnerSourceMetadata"] && repo["_InnerSourceMetadata"]["guidelines"] ? 100 : 0);
    // final score is a rounded value starting from 0
    iScore = Math.round(iScore - 1);
    // add score to metadata on the fly
    repo._InnerSourceMetadata = repo._InnerSourceMetadata || {};
    repo._InnerSourceMetadata.score = iScore;
    return iScore;
}
```

## Resulting Context

Projects can be sorted and presented by the InnerSource activity score to give a meaningful order in a list or portal. The activity score can be calculated on the fly or in a background job that evaluates all projects on a regular basis and stores a list of results.

> A crawler that regularly searches all InnerSource repositories (e.g. tagged with a certain label in an enterprise GitHub instance) can be a helpful solution as well. It provides a ranked list of projects that can be used as an input for tools like the [InnerSource Portal](innersource-portal.md), a search engine, or an interactive chat bot. 

Reviewer: What do you mean by "enterprise GitHub instance"? Same as "GitHub organization"?

Reviewer 2: The resulting context might also indicate the impact of not including the soft KPIs (e.g., the ranking might not be as meaningful in certain aspects, which might later point to another pattern that would work in the soft KPIs).

Author: reworked this part, an enterprise GitHub instance is a self-hosted GitHub instance inside the company, similar to a paid plan on GitHub where you can host private repos

Author: will leave out the soft KPI part here, i moved it to the solution and the caveats are described in forces

## Rationale

The InnerSource activity score is a simple calculation based on the GitHub API. It can be fully automated and easily adapted to new requirements. 

## Known Instances (optional)
Used in SAP's InnerSource project portal to define the default order of the InnerSource projects. It was first created in July 2020 and is fine-tuned and updated frequently ever since.

When proposed to InnerSourceCommons in July 2020, this pattern emerged.

## Status (optional until merging)

> First Draft: 30th August 2020  

Reviewer: I assume a month confusion here? :)

```suggestion
First Draft: 30th July 2020  
```
